### PR TITLE
added quick reply feature

### PR DESCRIPTION
**Pull request by abhishekchotaliya**

### CODE REVIEW

File: lib/twitter_dm_bot.js [hunk 1]
```diff
@@ -211,35 +211,73 @@ class TwitterDMBot extends BaseBot {
...
-    return new Promise((resolve, reject) => {
-      this.twit.post('direct_messages/new', rawMessage, (err, data) => {
-        if (err) {
-          reject(err);
-        }
-        resolve(data);
+  __sendMessage(rawMessage, send_options, quick_reply = false) {
+    console.log("rM");
+    console.log(util.inspect(rawMessage, false, null))
+
```
Comment (jdwuarin) [inline lib/twitter_dm_bot.js]: These console.logs should be removed

**Comment by abhishekchotaliya** [comment]
Please note that some changes are required in the botmaster repository to implement quick reply feature in botmaster-twitter-dm.

File: lib/twitter_dm_bot.js [hunk 2]
```diff
@@ -211,35 +211,73 @@ class TwitterDMBot extends BaseBot {
...
-    if (message.message.quick_replies) {
-      // Twitter DMs don't support buttons, so we'll just append the text with the buttons
-      // written in plain text and new lines.
-      const buttonsText = message.message.quick_replies.reduce((text, quickReply) => {
-        return `${text}\n${quickReply.title}`;
-      }, '');
+    if (message.quick_replies) {
+      // construct the quick replies for twitter platform
+
+      let newObj = {
```
Comment (jdwuarin) [inline lib/twitter_dm_bot.js]: Rename `newObj` to `quickReplyFormattedMessage` and should be const

**Comment by jdwuarin** [comment]
Thanks for this request. I've added a few comments above so that I can merge the code in. Should be like 5 minutes to do so.

Cheers

File: lib/twitter_dm_bot.js [hunk 3]
```diff
@@ -212,34 +212,85 @@ class TwitterDMBot extends BaseBot {
...
+    } else {
+      return new Promise((resolve, reject) => {
+        this.twit.post('direct_messages/new', rawMessage, (err, data) => {
+          if (err) {
+            reject(err);
+          }
+          resolve(data);
+        });
+      });
+    }
```
Comment (jdwuarin) [inline lib/twitter_dm_bot.js]: This code should look like this:

```js
return new Promise((resolve, reject) => {
  // if sending events like quick replies, add /events to url
  this.twit.post(`direct_messages${rawMessage.event ? '/events/' : '/'}new`, rawMessage, (err, data) => {
    if (err) {
      reject(err);
    }
    resolve(data);
  });
});
```

or if wanting to be more explicit:

```js
// if sending events like quick replies, add /events to url
const sendMessageUrl = rawMessage.event ? 'direct_messages/events/new' : 'direct_messages/new'
return new Promise((resolve, reject) => {
  this.twit.post(sendMessageUrl, rawMessage, (err, data) => {
    if (err) {
      reject(err);
    }
    resolve(data);
  });
});
```

File: lib/twitter_dm_bot.js [hunk 4]
```diff
@@ -212,34 +212,85 @@ class TwitterDMBot extends BaseBot {
...
-      const buttonsText = message.message.quick_replies.reduce((text, quickReply) => {
-        return `${text}\n${quickReply.title}`;
-      }, '');
-
-      formattedMessage.text += `\n(Please type in your choice)${buttonsText}`;
+      let quick_replies = [];
+      
+      // convert default quick reply structure to twitter supported structure
+      
+      let data = message.message.quick_replies;
```
Comment (jdwuarin) [inline lib/twitter_dm_bot.js]: `data` should be renamed to `quickReplies`. It should be a `const` as the object is never overwritten. You should be using the eslint config from the project here (found in the lib directory)

File: lib/twitter_dm_bot.js [hunk 5]
```diff
@@ -212,34 +212,85 @@ class TwitterDMBot extends BaseBot {
...
+      for (const d in data) {
+        let temp = {
+          label: data[d]["title"],
+          metadata: data[d]["title"]
+        }
+        quick_replies.push(temp)
+      }
```
Comment (jdwuarin) [inline lib/twitter_dm_bot.js]: this whole code should be changed to:

```js
const quickReplies = message.message.quick_replies.map(quickReply => ({
  label: quickReply['title'],
  metadata: quickReply['title']
}))
```

If you would rather use a loop, it should look like this:

```js
const quickReplies = []
for (const quickReply of message.message.quick_replies) {
  quickReplies.push({
    label: quickReply['title'],
    metadata: quickReply['title']
  })
}
```

File: lib/twitter_dm_bot.js [hunk 6]
```diff
@@ -212,34 +212,85 @@ class TwitterDMBot extends BaseBot {
...
+          type: "message_create",
+          message_create: {
+            target: {
+              recipient_id: formattedMessage.user_id,
+            },
+            message_data: {
+              text: message.message.text,
+              quick_reply: {
+                type: "options",
+                options: quick_replies
```
Comment (jdwuarin) [inline lib/twitter_dm_bot.js]: This should then use the `quickReplies` object

File: lib/twitter_dm_bot.js [hunk 7]
```diff
@@ -212,34 +212,85 @@ class TwitterDMBot extends BaseBot {
...
+      
+      let data = message.message.quick_replies;
+      for (const d in data) {
+        let temp = {
+          label: data[d]["title"],
+          metadata: data[d]["title"]
+        }
+        quick_replies.push(temp)
+      }
+      let quickReplyObject = {
```
Comment (jdwuarin) [inline lib/twitter_dm_bot.js]: this should be a `const` as it is not reassigned

**Comment by abhishekchotaliya** [comment]
Hi @jdwuarin,
I've made all the suggested changes. Hopefully, this time you can merge the branch!

**Comment by jdwuarin** [comment]
@abhishekchotaliya - looking good to me. If you can confirm that this all works for you with both quick replies and normal messages, I'll go ahead and merge it.

**Comment by abhishekchotaliya** [comment]
Yeah, it is working with both normal messages and quick replies.
Quick reply implementation is not breaking any existing functionalities.

**Pull request merged**